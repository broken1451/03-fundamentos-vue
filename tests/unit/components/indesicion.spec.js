import { shallowMount } from "@vue/test-utils";
import Indecision from "../../../src/components/indecision";

describe("Indesicion component", () => {
  let wrapper;
  let clgSpy;

  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () =>
        Promise.resolve({
          answer: "yes",
          forced: false,
          image: "https://yesno.wtf/assets/yes/2.gif",
        }),
    })
  );

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, "log");
    jest.clearAllMocks();
  });

  it("debe de hacer match con el snapshot / debe de ser igual a la fotografia ", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Escribir en el input no debe de disparar nada (console.log)", async () => {
    // vm es la instancia de vue
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

    const input = wrapper.find("input");
    await input.setValue("Hola q tal");

    // expect(clgSpy).toHaveBeenCalled();
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
    // console.log(wrapper.vm);
    // mockConstructor: es una funcion que esta siendo simulada para que yo pueda evaluar si ha sido llamada o no
  });

  it('Escribit el "?" debe de disparar el getAnswer ', async () => {
    const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");
    const input = wrapper.find("input");
    await input.setValue("Hola q tal?");
    expect(clgSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  it("Pruebas en metodo getAnswer", async () => {
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("Si!");
    // console.log(wrapper.vm.img);
    // console.log(wrapper.vm.answer);
  });

  it("Pruebas en metodo getAnswer - Fallo en el API", async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API is down"));
    await wrapper.vm.getAnswer();
    const img = wrapper.find("img");
    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe("No se pudo cargar del API");
  });
});
