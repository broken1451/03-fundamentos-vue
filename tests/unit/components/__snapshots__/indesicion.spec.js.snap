// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indesicion component debe de hacer match con el snapshot / debe de ser igual a la fotografia  1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="indecision-container"><input type="text" placeholder="Realiza una pregunta">
    <p>Recuerda terminar con un signo de interrogacion(?)</p>
    <!--v-if-->
  </div>
</div>
`;
